Add new target armv7-unknown-linux-uclibceabi (softfloat)

lancethepants opened the pull request:
This adds the new target `armv7-unknown-linux-uclibceabi (softfloat)`. It is of course similar to `armv7-unknown-linux-uclibceabihf (hardfloat)` which was just recently added to rust except that it is `softfloat`.

My interest lies in the Broadcom BCM4707/4708/BCM4709 family, notably found in some Netgear and Asus consumer routers. The armv7 Cortex-A9 cpus found in these devices do not have an fpu or NEON support.

With this patch I've been able to bootstrap rustc, std and host tools `(extended = true)` to run on the target device for native compilation, allowing the target to be used as a development platform.

With the recent addition of `armv7-unknown-linux-uclibceabihf (hardfloat)` it looks like many of the edge cases of using the uclibc c-library are getting worked out nicely. I've been able to compile some complex projects. Some patching still needed in some crates, but getting there for sure.  I think `armv7-unknown-linux-uclibceabi` is ready to be a tier 3 target.

I use a cross-toolchain from my project to bootstrap rust. 
https://github.com/lancethepants/tomatoware
The goal of this project is to create a native development environment with support for various languages.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nagisa (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

rust-highfive commented:
:warning: **Warning** :warning:

* These commits modify **compiler targets**. (See the [Target Tier Policy](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html).)

nagisa commented:
Please copy the list of tier requirements from [here](https://doc.rust-lang.org/rustc/target-tier-policy.html) and verify that each of the requirements are being satisfied.

Please also add a file representing this target to the [src/doc/rustc/src/platform-support](https://github.com/rust-lang/rust/tree/master/src/doc/rustc/src/platform-support) directory.

bors commented:
:umbrella: The latest upstream changes (presumably #93220) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lancethepants commented:
Rebase on Master

lancethepants commented:
Since submitting this pull request

Fixed up some uclibc issues in libc crate.
https://github.com/rust-lang/libc/pull/2615

uclibc support was added to nix crate.
https://github.com/nix-rust/nix/pull/1603

I also added file representing this target to the src/doc/rustc/src/platform-support directory.
https://github.com/rust-lang/rust/blob/b1039e94c13bf0149c6fc3e0b339693f378fc2ad/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5028617448?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b1039e94c13bf0149c6fc3e0b339693f378fc2ad/checks/5028617448/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 629 error codes
* highest error code: E0787
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:13: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:18: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:22: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:44: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:56: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:60: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:66: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:84: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:96: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:116: trailing whitespace
Found 0 error codes with no tests
Done!
* 363 features
some tidy checks failed
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5028617410?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b1039e94c13bf0149c6fc3e0b339693f378fc2ad/checks/5028617410/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 629 error codes
* highest error code: E0787
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:13: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:18: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:22: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:44: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:56: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:60: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:66: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:84: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:96: trailing whitespace
tidy error: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:116: trailing whitespace
Found 0 error codes with no tests
Done!
* 363 features
some tidy checks failed
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5029042937?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/63b062191862acd406f7890f3e85c44130d4f05e/checks/5029042937/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Set({doc::src/doc/edition-guide}) not skipped for "bootstrap::doc::EditionGuide" -- not in [src/tools/tidy]
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.15s
rustc/platform-support.html:332: broken link - `rustc/platform-support/armv7-unknown-linux-uclibceabi.html`
rustc/print.html:7064: broken link - `rustc/platform-support/armv7-unknown-linux-uclibceabi.html`
number of HTML files scanned: 32033
number of HTML redirects found: 9927
number of links checked: 2202659
number of links ignored due to external: 108200
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5029737765?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/9c7b3a4e234439e8e75a70e3def2dd41a7fec6cf/checks/5029737765/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Set({doc::src/doc/edition-guide}) not skipped for "bootstrap::doc::EditionGuide" -- not in [src/tools/tidy]
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.16s
rustc/platform-support.html:332: broken link - `rustc/platform-support/armv7-unknown-linux-uclibceabi.html`
rustc/print.html:7064: broken link - `rustc/platform-support/armv7-unknown-linux-uclibceabi.html`
number of HTML files scanned: 32033
number of HTML redirects found: 9927
number of links checked: 2202668
number of links ignored due to external: 108201
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5038600110?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/2a442a297c717ddc1096dc6473afc542d89be3c3/checks/5038600110/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
doc tests for: /checkout/src/doc/rustc/src/platform-support/aarch64-apple-ios-sim.md
doc tests for: /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md" "--test-args" ""

stdout ----

running 4 tests
running 4 tests
test /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Cross_Compilation (line 77) ... FAILED
test /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Test_with_QEMU (line 92) ... FAILED
test /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Cross_Compilation (line 72) ... FAILED
test /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Native_compilation (line 59) ... FAILED
failures:


---- /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Cross_Compilation (line 77) stdout ----
error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:78:51
  |
1 | CC=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-gcc \

error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:79:52
  |
  |
2 | CXX=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-g++ \

error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:80:50
  |
  |
3 | AR=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-ar \

error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:81:99
  |
  |
4 | CARGO_TARGET_ARMV7_UNKNOWN_LINUX_UCLIBCEABI_LINKER=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-gcc \

error[E0762]: unterminated character literal
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:82:55
  |
  |
5 | CARGO_TARGET_ARMV7_UNKNOWN_LINUX_UCLIBCEABI_RUSTFLAGS='-Clink-arg=-s -Clink-arg=-Wl,--dynamic-linker=/mmc/lib/ld-uClibc.so.1 -Clink-arg=-...

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0762`.
For more information about this error, try `rustc --explain E0762`.
Couldn't compile the test.
---- /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Test_with_QEMU (line 92) stdout ----
error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:93:51
  |
3 | CC=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-gcc \

error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:94:52
  |
  |
4 | CXX=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-g++ \

error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:95:50
  |
  |
5 | AR=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-ar \

error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:96:99
  |
  |
6 | CARGO_TARGET_ARMV7_UNKNOWN_LINUX_UCLIBCEABI_LINKER=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-gcc \

error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:97:146
  |
  |
7 | ...atoware/arm-soft-mmc/arm-tomatoware-linux-uclibcgnueabi/sysroot/" \

error: expected expression, found `/`
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:93:4
  |
  |
3 | CC=/opt/tomatoware/arm-soft-mmc/bin/arm-linux-gcc \
  |    ^ expected expression
error: aborting due to 6 previous errors

Couldn't compile the test.
Couldn't compile the test.
---- /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Cross_Compilation (line 72) stdout ----
error: unknown start of token: \
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:73:30
  |
3 | rustup toolchain link stage2 \


error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `toolchain`
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:73:8
  |
3 | rustup toolchain link stage2 \
  |        ^^^^^^^^^ expected one of 8 possible tokens
error: aborting due to 2 previous errors

Couldn't compile the test.
Couldn't compile the test.
---- /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Native_compilation (line 59) stdout ----
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `rust_1`
 --> /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:60:9
  |
3 | dpkg -i rust_1.xx.x-x_arm.deb
  |         ^^^^^^ expected one of 8 possible tokens
error: aborting due to previous error

Couldn't compile the test.


failures:
    /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Cross_Compilation (line 72)
    /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Cross_Compilation (line 77)
    /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Native_compilation (line 59)
    /checkout/src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md - Tier__3::Building_Rust_programs::Test_with_QEMU (line 92)
test result: FAILED. 0 passed; 4 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.04s


stderr ----
```

</details>

lancethepants commented:
@nagisa Seems the CI is trying to execute my code snippet examples from my platform-support page?

lancethepants commented:
So am I going to be limited as to what I can put on my platform-support page? I spent many hours creating and editing the page with code samples to be the most helpful for anyone wanting to use this target. Seems when it's creating the doc it's parsing everything on the page and doesn't like some elements of the code samples.

Urgau reviewed (commented):
The platform support pages are part of the `rustc` book and as part of this the automated tests wants to check that every code is working (compiling, runnable, ...) this is done via `rustdoc` who will analyze every markdown file and test every code block but `rustdoc` assume by default that if a code block doesn't have a decorator it's a Rust code but in you're case it's not. The fix is simple you just need to add the `text` decorator to tell `rustdoc` that the code blocks aren't rust code.

Urgau commented on src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:
````diff
@@ -0,0 +1,121 @@
+* `apt update && apt install rust`
+
+If you bootstrap rust on your own using the project above, it will create a .deb file that you then can install with
+```
````
```suggestion
```text
```

Urgau commented on src/doc/rustc/src/platform-support/armv7-unknown-linux-uclibceabi.md:
````diff
@@ -0,0 +1,121 @@
+    ${HOME}/rust-bootstrap-armv7-unknown-linux-uclibceabi/src/rust/rust/build/x86_64-unknown-linux-gnu/stage2
+    ```
+* Build with:
+    ```
````
```suggestion
    ```text
```

nagisa commented:
@bors r+ Thanks!

bors commented:
:pushpin: Commit 8c6f7fd5e1b6face1550249375743960cbde96f3 has been approved by `nagisa`

<!-- @bors r=nagisa 8c6f7fd5e1b6face1550249375743960cbde96f3 -->
<!-- homu: {"type":"Approved","sha":"8c6f7fd5e1b6face1550249375743960cbde96f3","approver":"nagisa"} -->
